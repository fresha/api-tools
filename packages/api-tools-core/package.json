{
  "name": "@fresha/api-tools-core",
  "version": "0.6.1",
  "description": "Core project functionality",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/surgeventures/api-tools.git",
    "directory": "packages/api-tools-core"
  },
  "main": "build/index.js",
  "types": "build/index.d.ts",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "build:watch": "npm run build -- --watch",
    "check": "run-s lint test",
    "check:fix": "run-s lint:fix test",
    "clean": "rimraf ./build",
    "eslint": "eslint ./src",
    "eslint:fix": "eslint ./src --fix",
    "lint": "run-s eslint typecheck",
    "lint:fix": "run-s eslint:fix typecheck",
    "prebuild": "npm run clean",
    "test": "jest",
    "typecheck": "tsc --noEmit"
  },
  "files": [
    "build/",
    "package.json",
    "LICENSE",
    "README.md"
  ],
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "API"
  ],
  "author": "Fresha Engineering",
  "contributors": [
    {
      "name": "Andriy Mykulyak",
      "email": "andriy@fresha.com",
      "url": "https://github.com/mykulyak"
    }
  ],
  "maintainers": [
    {
      "name": "Andriy Mykulyak",
      "email": "andriy@fresha.com",
      "url": "https://github.com/mykulyak"
    }
  ],
  "license": "MIT",
  "devDependencies": {
    "@fresha/eslint-config": "^0.3.0",
    "@fresha/jest-config": "^0.3.1",
    "@fresha/typescript-config": "^0.2.0",
    "typescript": "^5.0.2"
  }
}
