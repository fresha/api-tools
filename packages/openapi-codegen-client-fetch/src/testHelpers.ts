// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
