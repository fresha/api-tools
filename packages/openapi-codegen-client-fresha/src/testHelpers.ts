// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
