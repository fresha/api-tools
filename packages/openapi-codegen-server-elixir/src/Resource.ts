import console from 'console';

import type { Generator } from './Generator';

export class Resource {
  static makeModuleName(): string {
    return '';
  }

  static makeFileName(): string {
    return '';
  }

  readonly generator: Generator;

  constructor(generator: Generator) {
    this.generator = generator;
  }

  generateCode(): void {
    console.log(!!this.generator);
  }
}
