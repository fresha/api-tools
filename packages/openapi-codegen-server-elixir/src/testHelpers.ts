import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
