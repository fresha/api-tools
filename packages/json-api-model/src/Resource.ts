import { ITreeParent, TreeNode } from '@fresha/api-tools-core';

import { Attribute } from './Attribute';
import { Relationship } from './Relationship';

import type { IAttribute, IRegistry, IRelationship, IResource } from './types';
import type {
  AddBooleanPropertyOptions,
  AddNumberPropertyOptions,
  AddStringPropertyOptions,
  JSONSchema,
  JSONSchemaType,
} from '@fresha/json-schema-model';

type ResourceParent = ITreeParent<IRegistry> | Resource;

export class Resource extends TreeNode<IRegistry, ResourceParent> implements IResource {
  readonly type: string;
  readonly id: IAttribute;
  readonly attributes: Map<string, IAttribute>;
  readonly relationships: Map<string, IRelationship>;

  constructor(parent: ResourceParent, type: string) {
    super(parent);
    this.type = type;
    this.id = new Attribute(this, 'string');
    this.attributes = new Map<string, Attribute>();
    this.relationships = new Map<string, Relationship>();
  }

  addAttribute(name: string, type: 'boolean', options?: AddBooleanPropertyOptions): IAttribute;
  addAttribute(name: string, type: 'number', options?: AddNumberPropertyOptions): IAttribute;
  addAttribute(name: string, type: 'string', options?: AddStringPropertyOptions): IAttribute;
  addAttribute(name: string, type: 'object'): IAttribute;
  addAttribute(name: string, type: 'array'): IAttribute;
  addAttribute(
    name: string,
    type: JSONSchemaType,
    options?: AddBooleanPropertyOptions | AddNumberPropertyOptions | AddStringPropertyOptions,
  ): IAttribute {
    if (this.attributes.has(name)) {
      throw new Error(`Duplicate attribute ${name}`);
    }
    switch (type) {
      case 'boolean': {
        const result = new Attribute(this, type, options as AddBooleanPropertyOptions);
        this.attributes.set(name, result);
        return result;
      }
      case 'number': {
        const result = new Attribute(this, type, options as AddNumberPropertyOptions);
        this.attributes.set(name, result);
        return result;
      }
      case 'string': {
        const result = new Attribute(this, type, options as AddStringPropertyOptions);
        this.attributes.set(name, result);
        return result;
      }
      case 'object': {
        const result = new Attribute(this, type);
        this.attributes.set(name, result);
        return result;
      }
      case 'array': {
        const result = new Attribute(this, type);
        this.attributes.set(name, result);
        return result;
      }
      default:
        throw new Error(`Unsupported attribute type ${type}`);
    }
  }

  removeAttribute(name: string): void {
    this.attributes.delete(name);
  }

  clearAttributes(): void {
    this.attributes.clear();
  }

  addRelationship(name: string): IRelationship {
    if (this.relationships.has(name)) {
      throw new Error(`Duplicate relationship ${name}`);
    }
    const result = new Relationship();
    this.relationships.set(name, result);
    return result;
  }

  removeRelationship(name: string): void {
    this.relationships.delete(name);
  }

  clearRelationships(): void {
    this.relationships.clear();
  }

  // eslint-disable-next-line class-methods-use-this
  jsonSchema(): JSONSchema {
    throw new Error('Method not implemented.');
  }
}
