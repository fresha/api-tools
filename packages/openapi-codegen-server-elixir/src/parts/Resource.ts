import { snakeCase } from '@fresha/api-tools-core';

import type { Context } from './types';
import type { SourceFile } from '@fresha/code-morph-ex';
import type { ResourceModel } from '@fresha/json-api-model';

export class Resource {
  readonly context: Context;
  readonly moduleName: string;
  readonly sourceFile: SourceFile;
  readonly resource: ResourceModel;

  constructor(context: Context, moduleName: string, resource: ResourceModel) {
    this.context = context;
    this.moduleName = moduleName;
    this.sourceFile = this.context.project.createResourceFile(moduleName);
    this.resource = resource;
  }

  collectData(): void {
    this.context.logger.info(`Generating code for resource ${this.moduleName}`);
  }

  generateCode(): void {
    this.context.logger.info(`Generating code for resource ${this.moduleName}`);

    this.sourceFile.writeDefmodule(this.moduleName, () => {
      this.sourceFile.writeModuleAttr('resource_type', this.resource.type);
      this.sourceFile.writeUse('Jabbax.Document');
      this.maybeWriteBuildFunction();
      this.writeLinkFunction();
      this.maybeWriteLinkRelationshipsFunctions();
    });
  }

  protected maybeWriteBuildFunction(): void {
    if (this.resource.jsonSchema()) {
      for (const { otherResourceType } of this.resource.relationships.values()) {
        this.sourceFile.writeAlias(this.context.project.getResourceModuleName(otherResourceType));
      }
      this.sourceFile.writeFunction({
        name: 'build',
        params: ['config'],
        content: () => {
          this.sourceFile.writeStruct('Resource', () => {
            this.sourceFile.writePropertyAssignment('type', '@resource_type');
            this.sourceFile.writePropertyAssignment('id', 'config.id');
            this.sourceFile.writePropertyStruct('attributes', () => {
              for (const propName of this.resource.attributes.keys()) {
                const elixirPropName = snakeCase(propName);
                this.sourceFile.writeLine(`${elixirPropName}: config.${elixirPropName},`);
              }
            });

            if (this.resource.relationships.size) {
              this.sourceFile.writeLine('relationships:');
              this.sourceFile.writeIndented(() => {
                this.sourceFile.writeLine('%{}');
                for (const relName of this.resource.relationships.keys()) {
                  const elixirRelName = snakeCase(relName);
                  this.sourceFile.writeLine(
                    `|> link_relationship(:${elixirRelName}, config.${elixirRelName})`,
                  );
                }
              });
            }
          });
        },
      });
    }
  }

  protected maybeWriteLinkRelationshipsFunctions(): void {
    if (this.resource.relationships.size) {
      this.sourceFile.writeFunction({
        isPrivate: true,
        name: 'link_relationship',
        params: ['relationships', 'type', 'nil'],
        content: () => {
          this.sourceFile.writeLine(
            'Map.put(relationships, type, %Jabbax.Document.Relationship{data: nil})',
          );
        },
      });

      for (const [relName, { otherResourceType }] of this.resource.relationships) {
        const elixirRelName = snakeCase(relName);
        const moduleAlias = this.context.project.getResourceModuleAlias(otherResourceType);

        this.sourceFile.writeFunction({
          isPrivate: true,
          name: 'link_relationship',
          params: ['relationships', `:${elixirRelName}`, elixirRelName],
          content: () => {
            this.sourceFile.writeLine(
              `Map.put(relationships, :${elixirRelName}, ${moduleAlias}.link(${elixirRelName}))`,
            );
          },
        });
      }
    }
  }

  protected writeLinkFunction(): void {
    this.sourceFile.writeFunction({
      name: 'link',
      params: ['config'],
      content: () =>
        this.sourceFile.writeStruct('ResourceId', { type: '@resource_type', id: 'config.id' }),
    });
  }
}
