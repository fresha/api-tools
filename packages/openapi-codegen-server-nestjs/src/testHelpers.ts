import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
