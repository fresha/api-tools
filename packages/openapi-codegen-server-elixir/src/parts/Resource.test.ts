import { titleCase } from '@fresha/api-tools-core';
import {
  addResourceAttributes,
  addResourceRelationship,
  setResourceSchema,
} from '@fresha/openapi-codegen-utils';

import { createGenerator } from '../testHelpers';

import { Resource } from './Resource';

import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';

const makeResource = (
  resourceType: string,
): { resource: Resource; resourceSchema: SchemaModel } => {
  const generator = createGenerator();

  const resourceName = titleCase(resourceType);

  const resourceSchema = generator.context.openapi.components.setSchema(resourceName);
  setResourceSchema(resourceSchema, resourceType);

  const resource = new Resource(
    generator.context,
    generator.context.project.getResourceModuleName(resourceType),
    generator.context.registry.addResourceSchema(resourceSchema),
  );

  return { resource, resourceSchema };
};

test('for schema-less resources, only link/1 is generated', () => {
  const resourceType = 'schema-less';
  const generator = createGenerator();
  const resource = new Resource(
    generator.context,
    generator.context.project.getResourceModuleName(resourceType),
    generator.context.registry.addResourceSchema(resourceType),
  );

  resource.collectData();
  resource.generateCode();

  expect(resource.sourceFile.getText()).toMatchSnapshot();
});

test('empty resource', () => {
  const resourceType = 'micro-nano';
  const { resource } = makeResource(resourceType);
  resource.collectData();
  resource.generateCode();

  expect(resource.sourceFile.getText()).toMatchSnapshot();
});

test('happy path', () => {
  const generator = createGenerator();

  const resourceType = 'users';
  const resourceName = titleCase(resourceType);
  const resourceSchema = generator.context.openapi.components.setSchema(resourceName);

  // need to initialize resource schema before calling registry.parseResource
  setResourceSchema(resourceSchema, resourceType);
  addResourceAttributes(resourceSchema, {
    name: { type: 'string', required: true },
    age: 'integer',
    email: 'string',
  });
  addResourceRelationship(resourceSchema, 'organization', 'organizations');
  addResourceRelationship(resourceSchema, 'avatar', 'user-avatars');

  const resource = new Resource(
    generator.context,
    generator.context.project.getResourceModuleName(resourceType),
    generator.context.registry.addResourceSchema(resourceSchema),
  );

  resource.collectData();
  resource.generateCode();

  expect(resource.sourceFile.getText()).toMatchSnapshot();
});
