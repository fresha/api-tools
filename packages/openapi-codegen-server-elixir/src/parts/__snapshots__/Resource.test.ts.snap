// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty resource 1`] = `
"defmodule ApiToolsWeb.MicroNanoResource do
  @moduledoc false

  @resource_type "micro-nano"
  use Jabbax.Document

  def build(config) do
    %Resource{
      type: @resource_type,
      id: config.id,
      attributes: %{
      },
    }
  end

  def link(config) do
    %ResourceId{
      type: @resource_type,
      id: config.id,
    }
  end
end"
`;

exports[`for schema-less resources, only link/1 is generated 1`] = `
"defmodule ApiToolsWeb.SchemaLessResource do
  @moduledoc false

  @resource_type "schema-less"
  use Jabbax.Document

  def link(config) do
    %ResourceId{
      type: @resource_type,
      id: config.id,
    }
  end
end"
`;

exports[`happy path 1`] = `
"defmodule ApiToolsWeb.UsersResource do
  @moduledoc false

  @resource_type "users"
  use Jabbax.Document
  alias ApiToolsWeb.OrganizationsResource
  alias ApiToolsWeb.UserAvatarsResource

  def build(config) do
    %Resource{
      type: @resource_type,
      id: config.id,
      attributes: %{
        name: config.name,
        age: config.age,
        email: config.email,
      },
      relationships:
        %{}
        |> link_relationship(:organization, config.organization)
        |> link_relationship(:avatar, config.avatar)
    }
  end

  def link(config) do
    %ResourceId{
      type: @resource_type,
      id: config.id,
    }
  end

  defp link_relationship(relationships, type, nil) do
    Map.put(relationships, type, %Jabbax.Document.Relationship{data: nil})
  end

  defp link_relationship(relationships, :organization, organization) do
    Map.put(relationships, :organization, OrganizationsResource.link(organization))
  end

  defp link_relationship(relationships, :avatar, avatar) do
    Map.put(relationships, :avatar, UserAvatarsResource.link(avatar))
  end
end"
`;
